packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -0,0 +1,71 @@
+import {} from 'teleterm/ui/services/workspacesService/documentsService';
+import { routing } from 'teleterm/ui/uri';
+import AppContext from 'teleterm/ui/appContext';
+
+export async function retryWithRelogin<T>(
+  appContext: AppContext,
+  originatingDocumentUri: string,
+  resourceUri: string,
+  actionToRepeat: () => Promise<T>
+): Promise<T> {
+  let retryableErrorFromActionToRepeat: Error;
+  try {
+    return await actionToRepeat();
+  } catch (error) {
+    // TODO(ravicious): Replace this with actual check on metadata.
+    const isRetryable =
+      error instanceof Error &&
+      (error.message.includes('ssh: handshake failed') ||
+        error.message.includes('ssh: cert has expired'));
```

Comment:
I'll replace this with a check on metadata in another series of PRs. For now this is enough.